### PR TITLE
Dockerfile: use rustup and install 3 Rust versions

### PR DESCRIPTION
**Pull request by vrutkovs**
Fetch rust installer from `rustup.rs` and install 3 Rust versions (stable + 2 previous versions) so that tests would pass on several previous versions (would be added later)

TODO:
* [x] Sort out why CI is using previous container as a base 
  https://github.com/openshift/release/pull/4941 fixes that

### CODE REVIEW

File: dist/Dockerfile.build/Dockerfile
Comment (steveej) [inline dist/Dockerfile.build/Dockerfile:17]: do we need this even though `--default-toolchain` is defined above with this version? If this is required, I would favor to remove the version from the installation command.

**Review by steveej — COMMENTED** [review]
Two minor comments (actually the same but duplicate). I'd like to discuss if we can get rid of one of the Dockerfiles. I remember to have kept the symlink for the time of transitioning to a common path.

File: dist/Dockerfile.build/Dockerfile
Comment (vrutkovs) [reply dist/Dockerfile.build/Dockerfile:17]: I left `--default-toolchain 1.37.0` there so that whenever stable changes we'd still get 1.37 installed - and we'd have to update to new stable with a PR.

It can be removed if we commit to supporting "rolling stable" though

Comment (steveej) [reply dist/Dockerfile.build/Dockerfile:17]: Ideally we would install the most recent stable, parse the version value and also install and test against the previous 2 or 3 versions. A solution which doesn't require any hardcoding, neither in the install section nor in the tests, would be awesome.

Comment (vrutkovs) [reply dist/Dockerfile.build/Dockerfile:17]: Ah, got it, sure lets see if that can be nicely done

File: dist/Dockerfile.build/Dockerfile
Comment (steveej) [inline dist/Dockerfile.build/Dockerfile]: Fingers crossed Rust never releases version 2 or uses patch releases :smile:

**Review by steveej — COMMENTED** [review]
I think it is good enough for now. It would be possible to do something like

```
$ curl https://raw.githubusercontent.com/rust-lang/rust/master/RELEASES.md | grep '^Version' | head -n3 | awk '{ print $2 }' | xargs -L1 echo about to install
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  399k  100  399k    0     0   675k      0 --:--:-- --:--:-- --:--:--  675k
about to install 1.37.0
about to install 1.36.0
about to install 1.35.0
```

but I think it's not worth the hassle of adding the necessary special cases for patch releases which might be in between. This is definitely better than what we have now, and we can detect the available toolchains in the tests using `rustup toolchain list`.

/lgtm
/hold

I'll leave the final call to other reviewers or to yourself.

**Comment by steveeJ** [comment]
/lgtm case

As discussed in the weekly sync we'd like to install a fixed baseline version and also install the previous two versions.

**Comment by steveej** [comment]
/lgtm cancel

As discussed in the weekly sync we'd like to install a fixed baseline version and also install the previous two versions.

**Comment by vrutkovs** [comment]
Discussed on the meeting today: it would be confusing when the next Rust version goes stable, it gets pulled in during the build and if the tests break it would be unclear which change caused this. 

Instead of parsing the version we should set versions explicitly - also in order to have images reproducible on every commit

**Comment by vrutkovs** [comment]
/hold cancel

@steveeJ ready for review

File: dist/Dockerfile.build/Dockerfile
Comment (steveej) [inline dist/Dockerfile.build/Dockerfile]: This `mkdir` is leftover and not required for rustup. This made me wonder if `rustup` cleans up after itself, or if the installation files end up in the image.

**Review by steveej — CHANGES_REQUESTED** [review]
I hadn't caught this leftover before. sorry!

File: dist/Dockerfile.build/Dockerfile
Comment (vrutkovs) [reply dist/Dockerfile.build/Dockerfile]: Good catch, fixed

Comment (steveej) [reply dist/Dockerfile.build/Dockerfile]: Thanks. I checked the rustup script and it does clean up after itself so we're good.

**Review by steveej — APPROVED** [review]
Thanks!

/lgtm

**Pull request merged**
